---
title: "R Notebook"
output: html_notebook
---
```{r}
library(data.table)

dir_to_save_fig <- "out/"
dir.create(file.path(".", dir_to_save_fig), showWarnings = F, recursive = T)
```


```{r}

meta <- data.table(
  N = c(1:12),
  Custom_name = c("WT1_M", "WT1_IA", "WT1_MA", "WT1_MetSA", "WT2_M", "WT2_IA", "WT2_MA", "WT2_MetSA", "WT3_M", "WT3_IA", "WT3_MA", "WT3_MetSA" ),
  Conditions = c("M", "IA", "MA", "MetSA", "M", "IA", "MA", "MetSA", "M", "IA", "MA", "MetSA" ),
  Replica = c("1", "1", "1","1","2", "2", "2", "2",  "3", "3", "3", "3" ),
  Samples_Identifier = paste0("sample", c(1:12))
)

```



```{r}
library(data.table)
exp_dir <- '/storage1/fs1/martyomov/Active/collaborations/denisk/Tomas_Paulenda/macrophages_treated'

rownames(meta) <- meta$Samples_Identifier


count_files <- file.path(exp_dir, paste(meta$Samples_Identifier,"mouse", sep= '_'), "counts", "counts.txt")

all_counts <- list()

for (ind in 1:length(count_files)) {
  counts <- fread(count_files[[ind]], stringsAsFactors = F, header = TRUE)
  rownames(counts) <- counts$Geneid
  sample_id <- rownames(meta)[[ind]]
  names(counts) <- c(names(counts)[-ncol(counts)], sample_id)
  print('=================')
  print(paste('SAMPLE: ', sample_id))
  print('------------------')
  print( paste('total dimensions :',dim(counts)))

  all_counts[[sample_id]]<-counts[[sample_id]] 
}
all_counts <- as.data.frame(all_counts)
rn <-counts$Geneid
rownames(all_counts) <- rn

rowdata <- data.frame(gene_id = rn,gene_name =counts$gene_name, row.names = rn)
coldata <- meta

saveRDS(list(counts = all_counts, rowdata = rowdata, coldata = coldata), file = paste0(exp_dir, '/all_expr_data.rds'))
```


```{r}
rowdata <- rowdata[complete.cases(rowdata$gene_name),]
rowdata <- rowdata[rowdata$gene_name != "",]
all_counts <- all_counts[rownames(rowdata),]
all_counts_raw <- all_counts
rowdata_raw <- rowdata
coldata_raw <- coldata
# collapse repeated genes
all_counts <- as.matrix(all_counts)
rownames(all_counts) <- rowdata$gene_name
all_counts <- tapply(all_counts, list(row.names(all_counts)[row(all_counts)], colnames(all_counts)[col(all_counts)]), FUN = mean)
mode(all_counts) <- "integer"
all_counts <-  as.data.frame(all_counts)
all_counts <-  all_counts[,rownames(coldata)]

rowdata <- rowdata[!duplicated(rowdata$gene_name),]
rownames(rowdata) <- rowdata$gene_name
rowdata <- rowdata[rownames(all_counts),]

```



```{r}
pca_plots <- list()
```


```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(all_counts, colData = coldata, design=~1)
dds <- DESeq(dds)
rld <- rlogTransformation(dds)
library(PCAtools)
pcaData <- PCAtools::pca(counts(dds, normalized = TRUE), metadata=coldata) 
pca_plots$all_counts_normalized <- biplot(pcaData, colby = "Conditions", shape = "Replica", legendPosition = "right") + ggtitle("All samples all counts deseq normalized")
plot(pca_plots$all_counts_normalized)
pcaData <- PCAtools::pca(assay(rld), metadata=coldata) 
pca_plots$rlog_all <- biplot(pcaData, colby = "Conditions",  shape = "Replica", legendPosition = "right") + ggtitle("All samples all counts rlog")
plot(pca_plots$rlog_all)
```

## Same for top12k genes 

```{r}
library(data.table)
# bad_samples <- c("S24", "S31", "S27")
# coldata <- as.data.frame(coldata[!coldata$Samples_Identifier %in% bad_samples, ])
# rownames(coldata) <- coldata$Samples_Identifier
# all_counts <- all_counts[, ! colnames(all_counts) %in% bad_samples]


dds <- DESeqDataSetFromMatrix(all_counts, colData = coldata, design=~Conditions)
smallestGroupSize <- 3
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
dds <- DESeq(dds)
rld <- rlogTransformation(dds)
library(PCAtools)
pcaData <- PCAtools::pca(counts(dds, normalized = TRUE), metadata=coldata) 
pca_plots$clean_norm <- biplot(pcaData, colby = "Conditions", shape = "Replica", legendPosition = "right") + ggtitle("Without outliers all counts")
plot(pca_plots$clean_norm)
pcaData <- PCAtools::pca(assay(rld), metadata=coldata) 
pca_plots$clean_rlog <- biplot(pcaData, colby = "Conditions",  shape = "Replica", legendPosition = "right") + ggtitle ("Without outliers all counts rlog")

plot(pca_plots$clean_rlog )

means <- apply(assay(rld), 1, mean)
means <- sort(means, decreasing = TRUE)
means <- means[1:(ifelse(length(means) < 12000 ,yes = length(means), no = 12000))]
filt_vector <- rownames(rld) %in% names(means) 
rld_12k <- rld[filt_vector,]

pcaData <- PCAtools::pca(assay(rld_12k), metadata=coldata) 
pca_plots$rld_12k <- biplot(pcaData,  shape = "Replica", colby = "Conditions", legendPosition = "right", title = "Without outliers top12k genes rlog")
plot(pca_plots$rld_12k)

```

### Prepare object to be downloaded to phantasus
These are raw counts and collapsed (mean) by gene name counts

```{r}
library(Biobase)
library(dplyr)

get_object_for_phantasus <-  function(all_counts, rowdata, coldata, row_id='gene_name') {
  fData <- rowdata
  pData <- as.data.frame(coldata)
  rownames(fData) <- fData[[row_id]]
  rownames(pData) <- colnames(all_counts)
  metadata <- data.frame(labelDescription=names(pData),
                       row.names=names(pData))
  phenoData <- new("AnnotatedDataFrame", data=pData, varMetadata=metadata)
  metadata <- data.frame(labelDescription=names(fData),
                       row.names=names(fData))
  featureData <- new("AnnotatedDataFrame", data=fData, varMetadata=metadata)
  ed <- new ("MIAME",
          name="ItaProjectMacrophagesTreated",
          lab="",
          title="Tomas Paulenda itaconate project.Macrophages treated with MA, M, ITA, MetSA",
          contact="",
          pubMedIds="",
          url="",
          other=list())
  eset <- ExpressionSet(assayData = as.matrix(all_counts), phenoData = phenoData, featureData = featureData, experimentData = ed)
  ess <- list(chikv_cluster_cond=eset)
  return(ess)
}

# ess <- get_object_for_phantasus(all_counts, rowdata, coldata, row_id = 'gene_name')
# save(ess, file=paste0(exp_dir, '/denis_for_tom_treated_macro_6d18007c3a83b2e0_collapsed.rda'), version = 2)
# ess_raw <- get_object_for_phantasus(all_counts_raw, rowdata_raw, coldata_raw, row_id = 'gene_id')
# save(ess_raw, file=paste0(exp_dir, '/denis_for_tom_treated_macro6d18007c3a83b2e0_raw.rda'), version = 2)



```


## Just parse genes selected by Tom for future usage

```{r}
targetset_string <-  "Ifit2 Ifit3 Cmpk2 Isg15 Mx1 Isg20 Usp18 Gbp4 Gbp2 Rtp4 Batf2 Cxcl10 Ddx60 Herc6 Ifih1 Mov10 Ube2l6 Rsad2 Ifi35 Nmi Dhx58 Ifi44 Irf7 Sell Epsti1 Eif2ak2 Tap1 Parp12 Parp14 Parp9 Samd9l Stat1 Stat2 Irf1 Helz2 Sp110 Il15 Bst2 Uba7 Ifitm3 Trim21 Pnpt1 Adar Trafd1 Ogfr Tgtp2 Adgrg7 Tnfsf10 Ifit3b Ifi205 Gbp5 Ifit1bl1 Gm12250 Cd69 Mx2 Socs1 Gbp3 Gm18853 Iigp1 Enpp4 BC023105 Tmem171 AC125149.3 Pou3f1 Vcan Slfn1 Gbp7 Acod1 Phf11a Upp1 Angpt1 Gbp9 Gm43802 Lhx2 Oasl2 Pyhin1 Gm7592 Ddx58 Tarm1 Ifi203 Tlr3 Xaf1 AA467197 Gm12187 Clic5 Slfn9 Apol9b Trim30c Gm5431 Ccl12 Akap12 Mlkl Irgm1 Trex1 Klrg2 Timeless Misp Oas2 Mndal Dgat2 Oas3 Daxx Nt5c3 Setdb2 En2 Piwil4 Oas1a Olfr56 Il10 Slfn8 1600014C10Rik Rilpl1 Nlrc5 Phf11b Fam26f Gm2427 Zbp1 Aplnr I830077J02Rik Trim30d Gm13772 Gpsm2 Ifi47 Zufsp Clcn1 Phf11d Aldh1b1 Galnt3 Asap3 Mnda Irgm2 Slc25a22 Gm16340 Coch Trim30a Slc4a11 AF067063 Gm20482 Gm9134 Serpina3f Dll1 Tor3a Apol9a Pnp Asb13 Asb11 Uaca Psmb10 Dtx3l Znfx1 Gm12551 Ifi204 Il15ra Tnfaip8l3 Lgals9 Igtp Tpst1 Nectin4 Tmem67 Atp10a Gm8623 Ccnd2 Gnb4 Fcgr1 Papd7 Tgtp1 St3gal6 Sp100 Cnn3 Oasl1 Rbl1 Kif5c Sass6 Six1 Csprs 9430076C15Rik Nod1 Gm15433 Nap1l2 Gca Phf11c Gbp6 Srms Fgl2 Mitd1 Ccdc25 Dck Il18 Il18bp Pydc3 Gm16464 Hsh2d Gm4955 Glrp1 Acnat1 Slamf8 Gm2546 Arhgef37 Olfr93 Rpl10-ps3 Parp11 AC168977.1 Themis2 Cd274 Vwa3b Ccl8 Pttg1 Sp140 Glipr2 Gm10053 Ppa1 H2-T22 Gpr15 Gm12185 Rab19 Trim34a Casp7 BC147527 Gm5970 Unc80 Gm14219 Rnf135 Tlr11 Mthfr Il21r Ccnj Rgs14 Pcgf5 Hspa1b Mllt3 Ifit1 Usb1 Zbtb5 Rmdn3 Sptbn2 Parp3 Fcrl6 Cd86 Fbxw17 Arid5a Aida Sertad3 Gm16026 Tagap Cenpe Nsmaf Pml Lrp4 Fcgr4 Jak2 Arl4a Art3 Mafk Cd40 Samhd1 Arel1 Slamf9 Tiam1 March5 Tapbpl Gm8995 Bfsp1 Psme2b Dennd6b Pydc4 Bcl9 Ppm1k Col9a3 Rpgrip1l"


ifn_path <-  strsplit(targetset_string, "\\s+")[[1]]
```


# Diff expression using DeSeq Comparison of groups

```{r}
de_plots <- list()
dds_results <- list()
perform_DE <- function(all_counts, condition_1, condition_2) {
  selected_columns <- (coldata$Conditions == condition_1) |(coldata$Conditions == condition_2)

selected_coldata <- as.data.frame(coldata[selected_columns,])
selected_coldata$Conditions <- relevel(as.factor(selected_coldata$Conditions), ref= condition_1)
rownames(selected_coldata) <- selected_coldata$Samples_Identifier
selected_counts <- all_counts[rownames(rld_12k), selected_columns] # Took only 12k genes
 


dds_comparison <- DESeqDataSetFromMatrix(selected_counts, colData = selected_coldata, design=~Conditions)
dds_comparison <- estimateSizeFactors(dds_comparison)
dds_comparison <- DESeq(dds_comparison)


dds_res <- results(dds_comparison, contrast = c("Conditions",condition_2, condition_1 ), cooksCutoff = FALSE )
dds_res$gene_name <- rowdata[rownames(dds_res),'gene_name']
dds_res <- as.data.frame(dds_res)

dds_res_to_plot <- dds_res[rownames(rld_12k),]
dds_res_to_plot$ifn_path <- rownames(dds_res_to_plot) %in% ifn_path
dds_res_to_plot$prdx <- rownames(dds_res_to_plot) %in% c("Prdx5")
comparison_plot <- ggplot(dds_res_to_plot[complete.cases(dds_res_to_plot$padj),], aes(x=log2FoldChange, y=-log10(padj), color=ifn_path)) +
  geom_point() + theme_bw() + scale_color_manual(values=c("black", "red")) +
  geom_text_repel(data=dds_res_to_plot %>% dplyr::filter(padj < 0.05), aes(label=gene_name, color=NULL)) + ggtitle (paste0(condition_2," vs ", condition_1))

cur_shr <- lfcShrink(dds_comparison, coef= paste("Conditions", condition_2,"vs", condition_1, sep = "_"), type = "apeglm")

dds_res$log2FoldChange <- cur_shr$log2FoldChange
dds_res$lfcSE <- cur_shr$lfcSE

dds_res_to_plot <- dds_res[rownames(rld_12k),]
dds_res_to_plot$ifn_path <- rownames(dds_res_to_plot) %in% ifn_path
dds_res_to_plot$prdx <- rownames(dds_res_to_plot) %in% c("Prdx5")

comparison_plot_apeglm <- ggplot(dds_res_to_plot[complete.cases(dds_res_to_plot$padj),], aes(x=log2FoldChange, y=-log10(padj), color=ifn_path)) +
  geom_point() + theme_bw() + scale_color_manual(values=c("black", "red")) +
  geom_text_repel(data=dds_res_to_plot %>% dplyr::filter(padj < 0.05), aes(label=gene_name, color=NULL)) + ggtitle(paste0(condition_2," vs ", condition_1, "(apeglm)"))

return (list(dds_res = dds_res, plot1=comparison_plot, plot2=comparison_plot_apeglm))
}

```

## IA vs MA

```{r fig.width = 8, fig.height = 6}
condition_1 <-  "MA"
condition_2 <- "IA"
result <- perform_DE(all_counts, condition_1, condition_2) 
dds_results[[paste(condition_2, "vs", condition_1)]] <- result$dds_res
de_plots[[paste(condition_2, "vs", condition_1)]] <- result$plot1
result$plot1
result$plot2
```
## IA vs M

```{r fig.width = 8, fig.height = 6}
condition_1 <-  "M"
condition_2 <- "IA"
result <- perform_DE(all_counts, condition_1, condition_2) 
dds_results[[paste(condition_2, "vs", condition_1)]] <- result$dds_res
de_plots[[paste(condition_2, "vs", condition_1)]] <- result$plot1
result$plot1
result$plot2
```

## MA vs M

```{r fig.width = 8, fig.height = 6}
condition_1 <-  "M"
condition_2 <- "MA"
result <- perform_DE(all_counts, condition_1, condition_2) 
dds_results[[paste(condition_2, "vs", condition_1)]] <- result$dds_res
de_plots[[paste(condition_2, "vs", condition_1)]] <- result$plot1
result$plot1
result$plot2

```


## MetSA vs M


```{r fig.width = 8, fig.height = 6}
condition_1 <-  "M"
condition_2 <- "MetSA"
result <- perform_DE(all_counts, condition_1, condition_2) 
dds_results[[paste(condition_2, "vs", condition_1)]] <- result$dds_res
de_plots[[paste(condition_2, "vs", condition_1)]] <- result$plot1
result$plot1
result$plot2
```



# Alternative Diff expression using Limma 
this was looking better for Tom's perspective


## Methods
```{r}
library(limma)
library(edgeR)
limmaAnalysisSimpleImpl <- function(es, fieldValues, contrast){
    fieldValues <- replace(fieldValues, fieldValues == "", NA)

    es.copy <- es
    es.copy$Comparison <- fieldValues
    pData(es.copy)[,"Comparison"] <- as.factor(pData(es.copy)[,"Comparison"])
    pData(es.copy)[,"Comparison"] <- relevel(pData(es.copy)[,"Comparison"], ref = "Reference")
    fData(es.copy) <- data.frame(row.names = rownames(es.copy))
    es.copy <- es.copy[, !is.na(fieldValues)]

    # Getting rid of check NOTEs
    Comparison=ComparisonA=ComparisonB=NULL

    es.design <- stats::model.matrix(~0 + Comparison, data = pData(es.copy))
    es.copy <- voomNormalization( es.copy, filterByExp = T, designMatrix = es.design)
    
    
    fit <- lmFit(es.copy, es.design)

    A <- NULL; B <- NULL
    fit2 <- contrasts.fit(fit, makeContrasts(ComparisonTarget - ComparisonReference, levels = es.design))
    fit2 <- eBayes(fit2)
    de <- topTable(fit2, adjust.method = "BH", number = Inf)
    de <- de[row.names(fData(es.copy)), ]
    return(list(de=de, es=es.copy))
}
limmaAnalysisAdvancedImpl <- function(es, designData, contrast){
    ux_designMatrix <- getDesignMatrix(designData)

    es.copy <- es
    fData(es.copy) <- data.frame(row.names = rownames(es.copy))
    colnames(ux_designMatrix) <-  make.names(colnames(ux_designMatrix))
    target_level <- make.names(paste0(contrast[1],contrast[3]))
    reference_level <-  make.names(paste0(contrast[1],contrast[2]))


    fit <- lmFit(es.copy, ux_designMatrix)
    fit2 <- contrasts.fit(fit, makeContrasts(contrasts = paste(reference_level,target_level, sep = "-"), levels = ux_designMatrix))
    fit2 <- eBayes(fit2)
    de <- topTable(fit2, adjust.method = "BH", number = Inf)
    de <- de[row.names(fData(es.copy)), ]
    return(de)
}


#' Differential Expression analysis.
#'
#' \code{limmaAnalysis} performs differential expression analysis
#'     from limma package and returns a ProtoBuf-serialized resulting
#'     de-matrix.
#'
#' @param es ExpressionSet object. It should be normalized for
#'     more accurate analysis.
#'
#' @param fieldValues Vector of comparison values, mapping
#'     categories' names to columns/samples
#'
#' @param version name of the limma analysis implementation. Should be "One-factor design" or "Advanced design"
#'
#' @param contrast a character vector with exactly three elements: the name of a factor in the design formula, the name of the numerator level for the fold change, and the name of the denominator level for the fold change
#'
#' @param designData data.frame with design matrix
#'
#' @return Name of the file containing serialized de-matrix.
#'
#' @import Biobase
#' @import limma
#'
#' @examples
#' \dontrun{
#' data(es)
#' limmaAnalysis(es, fieldValues = c("A", "A", "A", "B", "B"))
#' }
#' @keywords internal
limmaAnalysis <- function (es, fieldValues, version = "One-factor design", contrast =  list('Comparison', 'Target', 'Reference'), designData = NULL) {
  fieldValues[fieldValues == contrast[3]] <- "Reference"
  fieldValues[fieldValues == contrast[2]] <- "Target"
  fieldValues <- replace(fieldValues, fieldValues == "", NA)
  de <- NULL
  contrast <- unlist(contrast)
  if (version == "One-factor design" ){
      simple_lima_result <- limmaAnalysisSimpleImpl(es, fieldValues, contrast)
      de <- simple_lima_result$de
      es <- simple_lima_result$es
  }
  if (version == "Advanced design"){
      de <- limmaAnalysisAdvancedImpl(es, designData, contrast)
  }
      deDf <- as.data.frame(de)
  toRemove <- intersect(colnames(fData(es)), colnames(deDf))
  fData(es)[, toRemove] <- NULL

  es$Comparison <- fieldValues
  fData(es) <- cbind(fData(es), deDf)
  assign("es", es, envir = parent.frame())

    # f <- tempfile(pattern = "de", tmpdir = getwd(), fileext = ".bin")
    # writeBin(protolite::serialize_pb(as.list(de)), f)
    # jsonlite::toJSON(f)
    return(es)
}



voomNormalization <- function(es,designMatrix, filterByExp = FALSE){
    es.copy <- es
    keep <- TRUE
    if (filterByExp){
        keep <- filterByExpr(exprs(es.copy), design = designMatrix)
    }
    es.copy <- es.copy[keep,]
    voom_counts <- voom(counts = exprs(es.copy), design=designMatrix)
    exprs(es.copy) <- voom_counts$E
    assayDataElement(es.copy, "weights") <- voom_counts$weights
    return( es.copy)
}



```

## Execution
```{r}

target_data <- all_counts
current_es <-  get_object_for_phantasus(target_data, rowdata, coldata, row_id = 'gene_name')[[1]]

de_results_lima = list()


condition_1 <-  "M"
condition_2 <- "IA"
de_result_lima <- limmaAnalysis(current_es, fieldValues = current_es$Conditions,  contrast = c("Conditions",condition_2, condition_1 ))
de_results_lima[[paste(condition_2, "vs", condition_1)]] <- de_result_lima

condition_1 <-  "MA"
condition_2 <- "IA"
de_result_lima <- limmaAnalysis(current_es, fieldValues = current_es$Conditions,  contrast = c("Conditions",condition_2, condition_1 ))
de_results_lima[[paste(condition_2, "vs", condition_1)]] <- de_result_lima

condition_1 <-  "M"
condition_2 <- "MA"
de_result_lima <- limmaAnalysis(current_es, fieldValues = current_es$Conditions,  contrast = c("Conditions",condition_2, condition_1 ))
de_results_lima[[paste(condition_2, "vs", condition_1)]] <- de_result_lima

condition_1 <-  "M"
condition_2 <- "MetSA"
de_result_lima <- limmaAnalysis(current_es, fieldValues = current_es$Conditions,  contrast = c("Conditions",condition_2, condition_1 ))
de_results_lima[[paste(condition_2, "vs", condition_1)]] <- de_result_lima
```

### Volcano plots 

```{r fig.width = 20, fig.height = 6}
library(data.table)
library(ggplot2)
library(Biobase)
library(plyr)
library(ggrastr)

de_lima_result_volcano_plots <- list()

for (cur_index in 1:length(de_results_lima)) {
  de_result_lima <- de_results_lima[[cur_index]]
  comparison <- names(de_results_lima)[[cur_index]]
  cur_pdt <- data.table(fData(de_result_lima))
  cur_pdt$gene_name <- rownames(fData(de_result_lima))
  cur_pdt[, mark := gene_name %in% ifn_path]
  xlims <- c(-max(abs(cur_pdt$logFC)), max(abs(cur_pdt$logFC)))
  ylims <- c(0, round_any(max(-log10(cur_pdt$adj.P.Val)), accuracy = 0.5, f = ceiling))
  ybreaks <- seq(from = ylims[1], to = ylims[2], by = 1)
  plot_title <- comparison
  cur_plot <-  ggplot(cur_pdt, aes(x=logFC, y=-log10(adj.P.Val))) +
      geom_point() +
      geom_point(cur_pdt[cur_pdt$mark,], mapping = aes(x=logFC, y=-log10(adj.P.Val)), color="red") +
      geom_hline(yintercept = -log10(0.05), linetype="dashed") +
      theme_classic(base_size = 14) +
      xlim(xlims) +
      xlab("log_FC") +
      ylab("-log10(adj P-value)") +
      coord_fixed() +
      scale_y_continuous(breaks=ybreaks, labels=paste(ybreaks)) +
      ggtitle(plot_title)
  
  de_lima_result_volcano_plots[[comparison]] <-  cur_plot

  
}

de_lima_result_volcano_plots
```
## Same plot style as it is in the paper

```{r}
de_lima_result_volcano_plots_styled <- list()

for (cur_index in 1:length(de_results_lima)) {
  de_result_lima <- de_results_lima[[cur_index]]
  comparison <- names(de_results_lima)[[cur_index]]
  cur_pdt <- data.table(fData(de_result_lima))
  cur_pdt$gene_name <- rownames(fData(de_result_lima))
  cur_pdt[, mark := gene_name %in% ifn_path]
  cur_pdt[, color := "1"]
  cur_pdt[adj.P.Val <= 0.05, color := "2"]
  cur_pdt[, color := factor(color, levels = c("1", "2"))]
  
  xlims <- c(-max(abs(cur_pdt$logFC)), max(abs(cur_pdt$logFC)))
  ylims <- c(0, round_any(max(-log10(cur_pdt$adj.P.Val)), accuracy = 0.5, f = ceiling))
  ybreaks <- seq(from = ylims[1], to = ylims[2], by = 1)
  plot_title <- comparison

  cur_plot <-  ggplot(cur_pdt, aes(x=logFC, y=-log10(adj.P.Val), color = color)) +
        ggrastr::rasterise(geom_point(), dpi=300) +
         ggrastr::rasterise(geom_point(cur_pdt[gene_name %in% ifn_path,], mapping = aes(x=logFC, y=-log10(adj.P.Val)),
                   color="#ba181b"), dpi=300) +
        geom_hline(yintercept = -log10(0.05), col="azure4", linetype=4) +
        theme_classic(base_size = 14) +
        xlim(xlims) +
        xlab("log_FC") +
        ylab("-log10(adj P-value)") +
        coord_fixed() +
        scale_y_continuous(breaks=ybreaks, labels=paste(ybreaks)) +
        scale_color_manual(values = c("#d3d3d3", "#0b090a")) +
        theme(legend.position = "none") +
        ggtitle(plot_title)
  
  
  out_path <- file.path(dir_to_save_fig, paste0(comparison, "_volcano_plot.png"))
  ggsave(out_path, dpi = 300, width = 6, height = 8)
  ggsave(gsub(".png", ".svg", out_path, fixed = TRUE), dpi = 300, width = 4, height = 6)
  
  de_lima_result_volcano_plots[[comparison]] <-  cur_plot

  
}
```





# Pathway analysis


## Get pathway lists

```{r}
library(msigdbr)
library(data.table)
library(fgsea)
library(gridExtra)
library(ggrastr)


kegg_sets <- msigdbr(species = "Mus musculus", category = "C2",  subcategory = "CP:KEGG")
react_sets <- msigdbr(species = "Mus musculus", category = "C2",  subcategory = "CP:REACTOME")
wiki_sets <-  msigdbr(species = "Mus musculus", category = "C2",  subcategory = "CP:WIKIPATHWAYS")
c2_sets <- rbind(kegg_sets, react_sets,wiki_sets )
c2_sets <- data.table(c2_sets)
c2_sets <- c2_sets[, .(sets = list(gene_symbol)), by = gs_name]
c2_gene_sets <- c2_sets$sets 
names(c2_gene_sets) <- c2_sets$gs_name

hall_sets <- msigdbr(species = "Mus musculus", category = "H")
hall_sets <- data.table(hall_sets)
hall_sets <- hall_sets[, .(sets = list(gene_symbol)), by = gs_name]
hall_gene_sets <- hall_sets$sets 
names(hall_gene_sets) <- hall_sets$gs_name

go_sets <- msigdbr(species = "Mus musculus", category = "C5", subcategory = "GO:BP" )
go_sets <- data.table(go_sets)
go_sets <- go_sets[, .(sets = list(gene_symbol)), by = gs_name]
go_gene_sets <- go_sets$sets 
names(go_gene_sets) <- go_sets$gs_name
```

## Plot selected pathways in PCA space

```{r}
target_data <- all_counts
current_es <-  get_object_for_phantasus(target_data, rowdata, coldata, row_id = 'gene_name')[[1]]
current_es <- exprs(current_es)
means <- apply(current_es, 1, mean)
means <- sort(means, decreasing = TRUE)
means <- means[1:(ifelse(length(means) < 12000 ,yes = length(means), no = 12000))]
filt_vector <- rownames(current_es) %in% names(means) 

current_es <- current_es[filt_vector,]
rowdata <-  rowdata[filt_vector,]
current_es <- log2(current_es + 1)


pcaData <- PCAtools::pca(current_es, metadata=coldata) 
new_pca_plot <- biplot(pcaData, colby = "Conditions",  legendPosition = "right") + ggtitle("Original PCA of log(x+1) data")
new_pca_plot
```
## GESECA to find candidate enriched paths
```{r}
library(fgsea)

current_pathways <- hall_sets$sets
names(current_pathways) <-  hall_sets$gs_name
set.seed(2)
gesecaRes <- geseca(current_pathways,current_es, minSize = 15, maxSize = 500)
plotGesecaTable(gesecaRes |> head(10), current_pathways, E=current_es)
```

## Generate new pca object directly
should also work with original pca_data$loadings

```{r}
library(ggrepel)

pathwaysToPlot <- c("HALLMARK_INTERFERON_ALPHA_RESPONSE")  # Result pathways to be plotted (choose above)

## Build PCA
E <- current_es
pdata <- coldata
fdata <- rowdata
colnames(E) <- es$title
E <- t(base::scale(t(E), center=TRUE, scale = FALSE))
pca <- prcomp(t(E))
explained <- (pca$sdev)^2 / sum(pca$sdev^2)
head(explained * 100)
xs <- sprintf("PC%s", seq_along(explained))
xlabs <- sprintf("%s (%.1f%%)", xs, explained * 100)
d <- cbind(as.data.frame(pca$x), pdata)

c1 <- 1; c2 <- 2

pcaPP <- ggplot(data=d, aes_string(x = "PC1", y = "PC2", color = "Conditions")) + 
    geom_point(size=4, 
               aes_string(x="PC1", y="PC2", fill="Conditions"), color='black', stroke =1, pch = 21) +  
  scale_fill_manual(values = c("#F37E80", "white",  "#35B34A", "#F1B77B")) +
    xlab(xlabs[c1]) + ylab(xlabs[c2]) 

## Project all paths to PCA space 
current_pathways <- lapply(current_pathways, intersect, y=rownames(E))
current_pathways <- lapply(current_pathways, unique)
current_pathways <- current_pathways[lengths(current_pathways) >= 5 & lengths(current_pathways) <= 250]

pathwayVectors <- lapply(current_pathways, function(pathway) {
    x <- setNames(rep(0, nrow(E)), rownames(E))
    x[pathway] <- 1
    x <- x/sqrt(sum(x))
    x
})
pathwayMatrix <- do.call(cbind, pathwayVectors)
pathwaysInPca <- t(t(pca$rotation) %*% pathwayMatrix)
pctVars <- colSums((t(E) %*% pathwayMatrix)) / mean(colSums(E))*100
pathwaysT <- as.data.table(pathwaysInPca[, c("PC1", "PC2", "PC3")],
                           keep.rownames = "pathway")
pathwaysT[, pctVar := pctVars]
pathwaysT[, label := sprintf("%s", pathway)]


pathwaysTT <- pathwaysT[pathway %in% pathwaysToPlot]


## Plot this PCA with projected pathways
linewidth <-  1
result_plot <- pcaPP  +
  geom_segment(data=pathwaysTT, aes(x=0, y=0, xend = PC1*120, yend = PC2*120),
              color="darkgrey", linewidth=1,
               arrow=arrow(length=unit(0.25, "cm"))) + 
  geom_text_repel(data=pathwaysTT,
            mapping=aes(label=gsub("HALLMARK_", "", label), x=PC1*80, y=PC2*120,),
            color="black") +

  #geom_text_repel(label=())+
  theme_classic(base_size = 12) + coord_fixed() + 
     theme(
    legend.position = "right",
     axis.line = element_line(color = "black", size = linewidth),
     axis.title = element_text(size = 14),
     panel.grid = element_blank(),
     plot.title = element_text(size = 12, face = "bold")
  )
out_path <- file.path(dir_to_save_fig,  "pca_with_selected_paths.png")
ggsave(out_path, dpi = 300, width = 8, height = 5)
ggsave(gsub(".png", ".svg", out_path, fixed = TRUE), dpi = 300, width = 8, height = 5)
result_plot
```

## FGSEA to calculate pathway enrichment pictures


```{r}
library(data.table)
padj_threshold <- 0.05
fgsea_results <- function(cur_table, gene_sets){
   inp_stats <- setNames(cur_table$stat, nm = cur_table$gene_name)
   fgsea_res <- fgsea(gene_sets, inp_stats)
   fgsea_res <- fgsea_res[padj <= padj_threshold]
   setorder(fgsea_res, padj)
   if(nrow(fgsea_res) == 0){
     print("!!! NO significant pathways !!!")
   }
   if(nrow(fgsea_res)>10){
        collapsedPathways <- collapsePathways(fgsea_res,
                                      gene_sets, inp_stats)
   mainPathways <- fgsea_res[pathway %in% collapsedPathways$mainPathways][
                         order(-NES), pathway]
   }else{
     mainPathways <- fgsea_res[order(-NES), pathway]
   }

   return( list(full_res = fgsea_res, collapsed =  mainPathways) )
}


library(ggplot2)
perform_PE <- function(de, gene_set, set_name, de_name, pe_dir){
         fgsea_res_path <-  file.path(pe_dir, paste(de_name, set_name,"rds", sep = "."))
         if(file.exists(fgsea_res_path)){
           message(paste(
             "!already exists!", fgsea_res_path, sep = "\n"
           ))
           return(TRUE)
         }
         de  <- de[!is.na(de$stat),]
         print(paste("start :", de_name ))
         res <- fgsea_results(cur_table = de, gene_sets = gene_set)
         saveRDS( object =  res, file = fgsea_res_path)
         return(TRUE)
}

pe_dir_deseq <- paste0(exp_dir, "/fgsea_deseq")
dir.create(pe_dir)
for (cur_index in 1:length(dds_results)) {
perform_PE(de  = dds_results[[cur_index]], gene_set = hall_gene_sets,  set_name = "hallmark", de_name = names(dds_results)[[cur_index]], pe_dir = pe_dir_deseq)
perform_PE(de  = dds_results[[cur_index]], gene_set = go_gene_sets,  set_name = "go", de_name = names(dds_results)[[cur_index]], pe_dir = pe_dir_deseq )
perform_PE(de  = dds_results[[cur_index]], gene_set = c2_gene_sets,  set_name = "kegg_c2", de_name = names(dds_results)[[cur_index]], pe_dir = pe_dir_deseq )
}


pe_dir_limma <- paste0(exp_dir, "/fgsea_limma")
dir.create(pe_dir)
for (cur_index in 1:length(de_results_lima)) {
 de_result_limma <-  fData(de_results_lima[[cur_index]])
 comparison <- names(de_results_lima)[[cur_index]]
 de_result_limma$gene_name <- rownames(de_result_limma)
 de_result_limma$stat <- de_result_limma$t
perform_PE(de  = de_result_limma, gene_set = hall_gene_sets,  set_name = "hallmark", de_name = comparison , pe_dir = pe_dir_limma)
perform_PE(de  = de_result_limma, gene_set = go_gene_sets,  set_name = "go", de_name = comparison, pe_dir = pe_dir_limma )
perform_PE(de  = de_result_limma, gene_set = c2_gene_sets,  set_name = "kegg_c2", de_name = comparison, pe_dir = pe_dir_limma )
}


```

## Read results

```{r}
main_pathways <- c()

read_pe_results <-  function(pe_dir) {
  fgsea_files <- list.files(pe_dir, pattern = ".rds")
  fgsea_objects <- data.table()

  for(fgsea_file in fgsea_files){
   fgsea_res <- readRDS(file = file.path(pe_dir,fgsea_file))
   main_pathways <- unique(c(main_pathways, fgsea_res$collapsed))
   fgsea_res <- fgsea_res$full_res
   comparison <- sub(x = fgsea_file, pattern = "(.+)\\.(.+)\\.rds", replacement = "\\1")
   gene_sets <- sub(x = fgsea_file, pattern = "(.+)\\.(.+)\\.rds", replacement = "\\2")
   fgsea_res$comparison <- comparison
   fgsea_res$set_name <- gene_sets
   fgsea_objects <- rbind(fgsea_objects, fgsea_res)
  }
  fgsea_objects$pathway <- as.factor(fgsea_objects$pathway)
  return(fgsea_objects)

}
fgsea_objects_deseq <-  read_pe_results(pe_dir_deseq)
fgsea_objects_limma <-  read_pe_results(pe_dir_limma)
```



```{r}
library(ggplot2)
library(ggrepel)
one_volcano <- function(fgsea_res,  comparison, set_name ) {
   ggplot(fgsea_res, aes(x=NES, y=-log10(padj), color=padj < 0.05)) +
  geom_point() + theme_bw() + scale_color_manual(values=c("black", "red")) +
  geom_text_repel(data=fgsea_res %>% dplyr::filter(padj < 0.05), aes(label=pathway, color=NULL)) + ggtitle(paste (comparison,set_name, sep = "."))
}


plot_all_volcano_plots <- function(fgsea_all) {
volcano_plots <- list()
comparisons <- unique(fgsea_all$comparison)
set_names <- unique(fgsea_all$set_name)
library(ggplot2)
library(ggrepel)

   for (cond in comparisons){
      for (sname in set_names){
         cur_res <- fgsea_all[comparison == cond & pathway %in% main_pathways & set_name == sname ,  ]
         clust_name <- cur_res$cluster_name[1]
         cur_plot <- one_volcano(cur_res, cond, sname)
         volcano_plots[[cond]][[paste(sname, sep = ".")]] <-  cur_plot
      }
   }
return(volcano_plots)
}

volcano_plots_deseq <- plot_all_volcano_plots(fgsea_all = fgsea_objects_deseq)
volcano_plots_limma <- plot_all_volcano_plots(fgsea_all = fgsea_objects_limma)

```


```{r fig.width = 15, fig.height = 8}
volcano_plots_limma
```



```{r}
library(stringr)
get_one_dot_plot <- function(fgsea_res,  gene_set){
   data <- fgsea_res[gene_set == gene_set, .(pathway,comparison, NES, padj)]
   print(paste("unique pathways: ",length(unique(data$pathway))))
   # plot: dot plot
   if (gene_set == "halmark"){
     data$pathway_name <- str_sub(data$pathway,1, 30)
   } else {
      data$pathway_name <- data$pathway
   }
   ggplot(data = data, aes(x = comparison, y = pathway_name, 
                        color = NES , size = -log10(padj))) + 
   geom_point() +
   scale_color_gradient(low = "blue", high = "red") +
   theme_bw() + 
   ylab("") + 
   xlab("") + 
   ggtitle( paste( gene_set, "pathways")) +  theme(axis.text.x = element_text(angle = 45, vjust = 0.5))
   }


generate_dot_plots <- function(fgsea_all){
    dot_plots <- list()
  
  for (geneset in unique(fgsea_all$set_name)){
         fgsea_all %>% filter(padj < 0.05 & set_name == geneset) %>% 
         arrange(padj) -> cur_fgsea
        cur_plot <- get_one_dot_plot(fgsea_res = cur_fgsea,gene_set = geneset)
        dot_plots[[geneset]] <- cur_plot
  }
    return(dot_plots)
}
  
dot_plots_deseq <- generate_dot_plots(fgsea_objects_deseq)
dot_plots_limma <- generate_dot_plots(fgsea_objects_limma)

```
```{r}
dot_plots_deseq[["hallmark"]]
out_path <- file.path(dir_to_save_fig,  "dot_plot_deseq.png")
ggsave(out_path, dpi = 300, width = 10, height = 9)
ggsave(gsub(".png", ".svg", out_path, fixed = TRUE), dpi = 300, width = 10, height = 8)

dot_plots_limma[["hallmark"]]

out_path <- file.path(dir_to_save_fig,  "dot_plot_limma.png")
ggsave(out_path, dpi = 300, width = 10, height = 9)
ggsave(gsub(".png", ".svg", out_path, fixed = TRUE), dpi = 300, width = 10, height = 8)
```



```{r}
dot_plots_limma[["hallmark"]]

```


```{r, fig.width=7, fig.height=4}

results <- lapply(1:length(dds_results), function(current_index) {
  dds_res <- dds_results[[current_index]]
  inp_stats <- setNames(dds_res$stat, nm = dds_res$gene_name)
  p <- plotEnrichment(ifn_path,
              inp_stats) + labs(title=paste(names(dds_results)[[current_index]], "IFN custom pathway"))
  return(p)
})

results
```







```{r}
means <- apply(current_es[, c(2,4,6,8,10, 12)], 1, mean)
rowdata$mean_expr <-  means


new_row_data <-  rowdata[order(rowdata$mean_expr, decreasing = T),]
current_es <- current_es[new_row_data$gene_name,]


data <-  current_es

selecte_genes_string <- "Ifi44 Ifit3b Ifi205 Ifit2 Isg20 Ifit3 Rsad2 Irf7 Zbp1 Isg15 Ifit1 Oas3 Stat2 Bst2 Stat1 Ifi203 Ifih1 Ifi35 Ifitm3 Cd86"
selected_genes <- ifn_path
selected_genes <- strsplit(selecte_genes_string, "\\s+")[[1]]
present_genes <- selected_genes[selected_genes %in% rownames(data)]
present_genes <-  head(present_genes, 30)
present_genes[7] <- "Stat1"
present_genes [present_genes == "Sell"] <- "Stat2"


to_plot <- data[present_genes,]
#to_plot <- remove_zero_rows(to_plot)
colnames(to_plot) <- coldata$Conditions
# Minmax scaling for visualization only
to_plot <-
  (as.data.frame(t(apply(to_plot, 1, function(x)
    (x - min(x)) / (max(x) - min(x))))))

```


```{r fig.height = 10, fig.width = 8}
library(ComplexHeatmap)
library(circlize)
nrows <- dim(to_plot)[[1]]
ncols <- dim(to_plot)[[2]]
factor_columns <-
  factor(coldata$Conditions, levels = c("M",         "MA", "IA",   "MetSA"))
colors_v <- c("#4088A6", "#2EB86E", "#6FD8A8", "#FF7F00", "#D43E4F", "#FFD700", "#BF7E40")


col_fun = colorRamp2(c(0, 0.5, 1), c("blue", "white", "red"))


marker_map <- Heatmap(
  to_plot,
  color = col_fun,
  name = "IFN genes",
  cluster_rows = F,
  cluster_columns = FALSE,
  show_column_names = FALSE,
  show_row_names = TRUE,
  show_heatmap_legend = T,
  use_raster = T,
  rect_gp = gpar(col = "gray", lwd = 1),
    column_split = factor_columns,

      heatmap_legend_param = list(
        title = "Gene expression", at = c(0, 1), 
        labels = c("row_min", "row_max"),
        direction = "horizontal",
         title_position ="topcenter",
          border = "black",
        legend_width = unit(6, "cm")
    ),
  row_title_gp = gpar(fontsize = 20),
  column_title_side  = "top",
)
filename <-  "heatmap.svg"


svg(filename, width = 6, height = 10)
draw(marker_map, heatmap_legend_side = "bottom")
dev.off()
marker_map
```

```{r}
ct_order <- c(
  1, 5, 9,  # M
  3, 7,  11,
    2, 6, 10,  # IA
  4,8,12  # metSA
)
n_markers <- 3
gaps_after <- 2
gaps_after <- rep(gaps_after, each = 4)
col_gaps <- gaps_after * n_markers




plt <- pheatmap::pheatmap(
  to_plot[,ct_order],
  cluster_rows =F,
  cluster_cols = F,
  color = colorRampPalette(c("blue", "white", "red"))(100),
  legend = T,
  gaps_col = c(3,6,9),
  angle_col = 315,
  fontsize_col = 10
)
```



```{r}


raster_pdf(file.path(plot_dir, "all_report.pdf"), width = 13, height = 8, res = 300)

for(plt in pca_plots){
  plot(plt)
}
for (plt in other_plots){
  plot(plt)
}
for(plt in de_plots){
  plot(plt)
}

plot(dot_plots[["hallmark"]])   
      
  
dev.off()
```
